File: native/src/visibility.cpp
#include "visibility.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <stdexcept>

namespace vector_engine {
namespace {

constexpr double kParallelEpsilon = 1e-9;
constexpr double kPointEpsilonSquared = 1e-10;

double cross(double ax, double ay, double bx, double by) {
    return ax * by - ay * bx;
}

bool ray_segment_hit(const Point2& origin, double dx, double dy,
                     const Segment2& segment, double& distance, Point2& point) {
    const double sx = segment.second.x - segment.first.x;
    const double sy = segment.second.y - segment.first.y;
    const double denominator = cross(dx, dy, sx, sy);
    if (std::abs(denominator) <= kParallelEpsilon) return false;
    const double qx = segment.first.x - origin.x;
    const double qy = segment.first.y - origin.y;
    const double ray_distance = cross(qx, qy, sx, sy) / denominator;
    const double along_segment = cross(qx, qy, dx, dy) / denominator;
    if (ray_distance < 0.0 || along_segment < -kParallelEpsilon ||
        along_segment > 1.0 + kParallelEpsilon) {
        return false;
    }
    distance = ray_distance;
    point = {origin.x + ray_distance * dx, origin.y + ray_distance * dy};
    return true;
}

bool same_point(const Point2& first, const Point2& second) {
    const double dx = first.x - second.x;
    const double dy = first.y - second.y;
    return dx * dx + dy * dy <= kPointEpsilonSquared;
}

}  // namespace

VisibilityOutput visibility_polygon(const Point2& light,
                                    const std::vector<Segment2>& segments,
                                    double angle_epsilon) {
    if (!(angle_epsilon > 0.0)) {
        throw std::invalid_argument("angle_epsilon must be positive");
    }
    std::vector<double> angles;
    angles.reserve(segments.size() * 6);
    for (const auto& segment : segments) {
        for (const Point2 point : {segment.first, segment.second}) {
            const double angle = std::atan2(point.y - light.y, point.x - light.x);
            angles.push_back(angle - angle_epsilon);
            angles.push_back(angle);
            angles.push_back(angle + angle_epsilon);
        }
    }

    VisibilityOutput output;
    std::vector<VisibilityRay> hits;
    hits.reserve(angles.size());
    for (const double angle : angles) {
        const double dx = std::cos(angle);
        const double dy = std::sin(angle);
        double nearest = std::numeric_limits<double>::infinity();
        Point2 nearest_point{};
        std::size_t nearest_index = 0;
        bool found = false;
        for (std::size_t index = 0; index < segments.size(); ++index) {
            ++output.intersection_tests;
            double distance = 0.0;
            Point2 point{};
            if (ray_segment_hit(light, dx, dy, segments[index], distance, point) &&
                distance < nearest) {
                nearest = distance;
                nearest_point = point;
                nearest_index = index;
                found = true;
            }
        }
        if (found) hits.push_back({angle, nearest_point, nearest, nearest_index});
    }
    std::sort(hits.begin(), hits.end(), [](const auto& first, const auto& second) {
        return first.angle < second.angle;
    });
    for (const auto& hit : hits) {
        if (!output.rays.empty() && same_point(output.rays.back().point, hit.point)) {
            continue;
        }
        output.rays.push_back(hit);
    }
    if (output.rays.size() > 1 &&
        same_point(output.rays.front().point, output.rays.back().point)) {
        output.rays.pop_back();
    }
    output.polygon.reserve(output.rays.size());
    for (const auto& ray : output.rays) output.polygon.push_back(ray.point);
    return output;
}

}  // namespace vector_engine

File: src/core/visibility.py
"""Reference 2D visibility-polygon algorithm using ray/segment intersections."""

from dataclasses import dataclass
import math


PARALLEL_EPSILON = 1e-9
POINT_EPSILON = 1e-5


@dataclass(frozen=True)
class VisibilityRay:
    angle: float
    point: tuple
    distance: float
    segment_index: int


@dataclass(frozen=True)
class VisibilityResult:
    polygon: tuple
    rays: tuple
    segment_count: int
    intersection_tests: int
    backend: str = "Python reference"
    error: str = ""


def ray_segment_intersection(origin, direction, segment,
                             parallel_epsilon=PARALLEL_EPSILON):
    """Return (distance, point) for the forward ray hit, otherwise None."""
    ox, oy = map(float, origin)
    dx, dy = map(float, direction)
    (ax, ay), (bx, by) = segment
    sx, sy = float(bx) - float(ax), float(by) - float(ay)
    denominator = dx * sy - dy * sx
    if abs(denominator) <= parallel_epsilon:
        return None
    qx, qy = float(ax) - ox, float(ay) - oy
    distance = (qx * sy - qy * sx) / denominator
    along = (qx * dy - qy * dx) / denominator
    if distance < 0.0 or along < -parallel_epsilon or along > 1.0 + parallel_epsilon:
        return None
    return distance, (ox + distance * dx, oy + distance * dy)


def visibility_polygon(light, segments, angle_epsilon=1e-5):
    """Cast three rays around every segment endpoint and retain nearest hits."""
    light = (float(light[0]), float(light[1]))
    segments = tuple(((float(a[0]), float(a[1])),
                      (float(b[0]), float(b[1]))) for a, b in segments)
    if angle_epsilon <= 0:
        raise ValueError("angle_epsilon must be positive")
    angles = []
    for first, second in segments:
        for point in (first, second):
            base = math.atan2(point[1] - light[1], point[0] - light[0])
            angles.extend((base - angle_epsilon, base, base + angle_epsilon))
    hits = []
    tests = 0
    for angle in angles:
        direction = (math.cos(angle), math.sin(angle))
        nearest = None
        nearest_index = -1
        for index, segment in enumerate(segments):
            tests += 1
            hit = ray_segment_intersection(light, direction, segment)
            if hit is not None and (nearest is None or hit[0] < nearest[0]):
                nearest, nearest_index = hit, index
        if nearest is not None:
            hits.append(VisibilityRay(angle, nearest[1], nearest[0], nearest_index))
    hits.sort(key=lambda item: item.angle)
    filtered = []
    threshold_sq = POINT_EPSILON * POINT_EPSILON
    for hit in hits:
        if filtered:
            dx = hit.point[0] - filtered[-1].point[0]
            dy = hit.point[1] - filtered[-1].point[1]
            if dx * dx + dy * dy <= threshold_sq:
                continue
        filtered.append(hit)
    if len(filtered) > 1:
        dx = filtered[0].point[0] - filtered[-1].point[0]
        dy = filtered[0].point[1] - filtered[-1].point[1]
        if dx * dx + dy * dy <= threshold_sq:
            filtered.pop()
    return VisibilityResult(tuple(hit.point for hit in filtered), tuple(filtered),
                            len(segments), tests)
